llama-2-7b/code/web_example_text_completion.py

```diff
@@ -1,68 +1,67 @@
 
-    def append(self, new_prompt_tokens): 
+    def append(self, max_gen_len, new_prompt_tokens):
         available_tokens = self.max_seq_len - len(new_prompt_tokens)
-        if available_tokens < 0: 
+        if available_tokens < 0:
             print("User input exceeds the maximum token length")
-        
+
         # Account for tokens required for model output
-        available_tokens -= self.max_gen_len
+        available_tokens -= max_gen_len
 
         if available_tokens > len(self.token_history):
             self.token_history.extend(new_prompt_tokens)
             return self.token_history
-                
+
         # If the total tokens exceed the allowed limit, remove the earliest tokens
         global generator
-        print(generator.tokenizer.decode(self.token_history[-available_tokens:]), "context")
         self.token_history = self.token_history[-available_tokens:] + new_prompt_tokens
-        assert len(self.token_history) == available_tokens
+        assert len(self.token_history) == self.max_seq_len - max_gen_len
         return self.token_history
 
 # Default values for the generator
-default_values = {
+gen_params = {
     'ckpt_dir': 'weights/',
     'tokenizer_path': 'tokenizer.model',
     'max_seq_len': 128,
-    'max_batch_size': 4
+    'max_batch_size': 4,
+    'model_parallel_size': int(os.environ.get("WORLD_SIZE", 1)),
 }
 
 generator = Llama.build(
-    ckpt_dir=default_values['ckpt_dir'],
-    tokenizer_path=default_values['tokenizer_path'],
-    max_seq_len=default_values['max_seq_len'],
-    max_batch_size=default_values['max_batch_size']
+    ckpt_dir=gen_params['ckpt_dir'],
+    tokenizer_path=gen_params['tokenizer_path'],
+    max_seq_len=gen_params['max_seq_len'],
+    max_batch_size=gen_params['max_batch_size'],
 )
 
-window = SlidingWindow(max_seq_len=default_values['max_seq_len'])
+window = SlidingWindow(max_seq_len=gen_params['max_seq_len'])
 
 @app.route('/configure', methods=['POST'])
 def configure_generator():
     global generator
 
-    ckpt_dir = request.json.get('ckpt_dir', default_values['ckpt_dir'])
-    tokenizer_path = request.json.get('tokenizer_path', default_values['tokenizer_path'])
-    max_seq_len = int(request.json.get('max_seq_len', default_values['max_seq_len']))
-    max_batch_size = int(request.json.get('max_batch_size', default_values['max_batch_size']))
+    for key, value in gen_params.items():
+        gen_params[key] = request.json.get(key, value)
 
     generator = Llama.build(
-        ckpt_dir=ckpt_dir,
-        tokenizer_path=tokenizer_path,
-        max_seq_len=max_seq_len,
-        max_batch_size=max_batch_size,
+        ckpt_dir=gen_params['ckpt_dir'],
+        tokenizer_path=gen_params['tokenizer_path'],
+        max_seq_len=gen_params['max_seq_len'],
+        max_batch_size=gen_params['max_batch_size'],
+        # Reinitializing generator requires this additional param
+        model_parallel_size=gen_params['model_parallel_size']
```

Comment:
model_parallel_size, where is variable used?

Reply:
model_parallel_size specifies # of processes torch uses for training. It is used here: https://github.com/facebookresearch/llama/blob/main/llama/generation.py#L58-L81
User needs to set this as an env variable "WORLD_SIZE"